File: axpy/axpy_omp.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

// Inclusions 
#include <stdlib.h>
#include <iostream>
#include <iomanip>
#include <cmath>
#include "get_time.h"


// Example routine to perform some simple vector linear combinations. 
int main(int argc, char* argv[]) {

  // ensure that an argument was passed in
  if (argc < 2) {
    std::cerr << "Error: function requires one argument (vector length)\n";
    return 1;
  }

  // set n as the input argument, and ensure it's positive
  int n = atoi(argv[1]);
  if (n < 1) {
    std::cerr << "Error: vector length " << n << " must be greater than 0\n";
    return 1;
  }

  // allocate the vectors 
  double *x = new double[n];
  double *y = new double[n];
  double *z = new double[n];

  // start timer 
  double stime = get_time();

  // start OpenMP parallelism 
  double a = -3.0;
  double zmax = -1.0e300;
  #pragma omp parallel default(shared)
  {

    // output parallelism information 
#ifdef _OPENMP
    #pragma omp single
    std::cout << " starting OpenMP with " << omp_get_num_threads() << " processes\n";
#endif

    // initialize x and y 
    #pragma omp for
    for (int i=0; i<n; i++) {
      x[i] = exp(2.0*(i+1)/n);
      y[i] = 1.0*(n-1)/n;
    }

    // perform linear combinations, accumulate local maximum
    double zmaxl = -1.0e300;
    #pragma omp for 
    for (int i=0; i<n; i++) {
      z[i] = a*x[i] + y[i];
      x[i] = y[i]/a - z[i];
      y[i] = x[i]*y[i]/n;
      zmaxl = (zmaxl > z[i]) ? zmaxl : z[i];
    }

    // combine maximum values for each thread to global max 
    #pragma omp critical(overall_max)
    { zmax = (zmax > zmaxl) ? zmax : zmaxl; }

  } // end omp parallel 

  // output maximum value in z 
  std::cout << "  max(z) = " << std::setprecision(16) << zmax << std::endl;

  // stop timer 
  double ftime = get_time();

  // output total time 
  std::cout << " runtime = " << std::setprecision(16) << ftime - stime << std::endl;

  // free vectors 
  delete[] x;
  delete[] y;
  delete[] z;

  return 0;
} // end main 

File: axpy/axpy_omp.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU Mathematics
! Math 4370 / 6370
!=================================================================


program axpy
  !-----------------------------------------------------------------
  ! Description: 
  !    Performs some simple vector linear combinations.
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer, parameter :: MAX_INTEGER_DIGITS = 10
  integer :: n, i, numprocs
  character(MAX_INTEGER_DIGITS) :: n_string
  double precision :: a, stime, ftime, zmax, zmaxl
  double precision, allocatable :: x(:), y(:), z(:)
  double precision, external :: omp_get_wtime
  integer, external :: omp_get_num_threads

  !======= Internals ============

  ! get n from the command line
  call getarg(1, n_string)

  ! ensure that an argument was passed in
  if ( n_string == '' ) then
     stop 'Error: function requires one argument (vector length)'
  endif
  
  ! convert n_string to integer
  read (n_string, *) n

  ! allocate arrays
  allocate(x(n), y(n), z(n))

  ! start timer
  call get_time(stime)
  !$ stime = omp_get_wtime()    ! replace stime with OpenMP version (if available)

  ! start OpenMP parallelism
  a = -3.d0
  zmax = -1.0d-300
  !$omp parallel default(shared) private(i,zmaxl)

  ! ouptut parallelism information (only if OpenMP is enabled)
  !$ numprocs = omp_get_num_threads();
  !$omp single
  !$ print '(A,i2,A)',' starting OpenMP with ',numprocs,' processes'
  !$omp end single

  ! initialize x and y
  !$omp do
  do i=1,n
     x(i) = exp(2.d0*i/n)
     y(i) = 1.d0*(n-1)/n
  end do
  !$omp end do

  ! perform linear combination
  zmaxl = -1.0d300
  !$omp do
  do i=1,n
     z(i) = a*x(i) + y(i)
     x(i) = y(i)/a - z(i)
     y(i) = x(i)*y(i)/n
     zmaxl = max(zmaxl, z(i))
  end do
  !$omp end do

  ! combine maximum values for each thread to global max
  !$omp critical(overall_max)
  zmax = max(zmax, zmaxl)
  !$omp end critical(overall_max)

  ! end parallel region
  !$omp end parallel

  ! output maximum value in z
  write(*,*) '  max(z) =',maxval(z)

  ! stop timer
  call get_time(ftime)
  !$ ftime = omp_get_wtime()

  ! output total time
  write(*,*) ' runtime =',ftime-stime
  
  ! deallocate arrays
  deallocate(x, y, z)


end program axpy
!=================================================================
